core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java

```diff
@@ -95,30 +107,84 @@ public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand
 
    @Override
    public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) throws Throwable {
+      Map<Object, Object> originalMap = command.getMap();
+      ConsistentHash ch = dm.getConsistentHash();
+      Address localAddress = rpcManager.getAddress();
       if (ctx.isOriginLocal()) {
-         Set<Address> primaryOwners = new HashSet<Address>(command.getAffectedKeys().size());
-         for (Object k : command.getAffectedKeys()) {
-            primaryOwners.add(cdl.getPrimaryOwner(k));
+         List<NotifyingFuture<Object>> futures = new ArrayList<NotifyingFuture<Object>>(
+               rpcManager.getMembers().size() - 1);
+         // TODO: if async we don't need to do futures...
+         RpcOptions options = rpcManager.getDefaultRpcOptions(isSynchronous(command));
+         for (Address member : rpcManager.getMembers()) {
```

Comment:
It checks the segments on line 127 below and only sends it if there are some for that member.  In this case it will only send to either 1 or 2 owners based on where the segments are for those keys.